content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -579,41 +657,919 @@ sending values over the wire, the code should query the size of the
+You can, however, override the compiler's alignment choice using an 
+attribute definition clause or the :ada:`Alignment` aspect. You can do 
+so on types other than record types, but specifying it on record types 
+is typical. Here's our example record type with the alignment specified 
+via the aspect: 
+
+.. code-block:: ada
+
+   type My_Int is range 1 .. 10;
+   
+   subtype S is Integer range 1 .. 10;
+   
+   type R is record
+      M : My_Int;
+      X : S;
+      B : Boolean;
+      C : Character;
+   end record with
+      Alignment => 1;
+
+Alignment values are in terms of storage elements. The effect of the 
+aspect or attribute clause is to ensure that the starting address of the 
+memory allocated to objects of the type will be a multiple of the 
+specified value. 
+
+In fact, whenever we specify a record type layout we really should also 
+specify the record type's alignment, even though doing so is optional. 
+Why? The alignment makes a difference in the overall record object's 
+size. We've seen that already, with the padding bytes: the compiler will 
+respect the alignment requirements of the components, and may add 
+padding bytes within the record and also at the end to ensure components 
+start at addresses compatible with their alignment requirements. The 
+alignment also affects the size allocated to the record type even when 
+the components are already aligned. As a result the overall size could 
+be larger than we want for the sake of space. Additionally, when we pass 
+such objects to code written in other languages, we want to ensure that 
+the starting address of these objects is aligned as the external code 
+expects. The compiler might not chose that required alignment by 
+default. 
+
+Specifying alignment for record types is so useful that in the first 
+version of Ada there was no syntax to specify alignment for anything 
+other than record types (via the obsolete "at mod" clause on record 
+representation clauses). 
+
+For that reason GNAT provides a pragma named Optimize_Alignment. This is 
+a configuration pragma that affects the compiler's choice of default 
+alignments where no alignment is explicitly specified. There is a 
+time/space trade-off in the selection of these values, as we've seen. 
+The normal choice tries to balance these two characteristics, but with 
+an argument to the pragma you can give more weight to one or the other. 
+The best approach is to specify the alignments explicitly, per type, for 
+those that require specific alignment values. The pragma has the nice 
+property of giving general guidance to the compiler for what should be 
+done for the other types and objects not explicitly specified. 
+
+Now let's look into the details. We'll use a case study for this 
+purpose, including specifying sizes as well as alignments. 
+
+The code for the case study is as follows. It uses Size clauses to 
+specify the Sizes, instead of the Size aspect, just to emphasize that 
```

Comment:
```suggestion
The code for the case study is as follows. It uses :ada:`Size` clauses to 
specify the :ada:`Size`s, instead of the :ada:`Size` aspect, just to emphasize that 
```